### DISCUSSION TITLE
Possible to know when `x-show`/`x-transition` have completed showing or hiding the element?

### DISCUSSION THREAD

**Discussion by nikitaourazbaev** [alpinejs/alpine#3643] [1. Help] [answered]
Is it possible to be notified or run a callback when `x-show` has completed showing or hiding the element? 

I’m looking for a definitive way to know when something has been shown or hidden, regardless of whether transitions are used. Something that covers the cases of `x-show` with or without `x-transition`, and when `x-transition` is used, when a duration is set either with CSS classes or with the duration modifier. Ideally I would like to cover all of these without having to check whether a transition is being used.

Does Alpine know when the transition is complete? And is it possible to hook into that somehow?

**Comment by ekwoka** [comment, marked as answer]
When a transition happens, `transitionend` event is dispatched, this is native behavior. It won't happen when there is no transition.

You can use some code to tap into element internals, and, after x-show/x-transition is registered, wrap `el._x_doShow` in another function that also does whatever you want and put that in it's place.

Is there a reason why you'd want to hook into the action in the dom and not onto the values themselves?

There's a few different ways to do the general idea of knowing when the show evaluates, but which makes sense depends on why you want to do it in the first place.

**Reply by nikitaourazbaev** [reply]
Yes, I think overriding `_x_doHide` would work in my case, thank you!

I have a modals system managed via an Alpine store and need to run some teardown code once a modal is no longer shown. Some modals have transitions and others don't but the method closing them is the same. Currently I’m checking whether the `x-transition` attribute is present and either running the code immediately or on `transitionend`. 

However:

1. Ideally my closing method wouldn’t need to know if there’s a transition, only when the modal is fully closed and no longer visible
2. Looking at the source code of `x-transition` there’s some sophistication happening around deciding when a transition is finished, and I’d like to tap into that
3. I’ve run into browser compatibility issues in the past using `transitionend`, not sure if those are still relevant

**Reply by ekwoka** [reply]
hmm, the simplest way is if you were using `x-if` you could just put the teardown in a `destroy` key in the data context, it's like the opposite of `init`.

But x-if doesn't work with x-transition.

🤔 

Is there a real issue if you do the teardown logic just upon the variable becoming falsy? Does it matter if it hasn't fully transitioned out yet? Like what is it, just reseting the state? Or even just run it on it becoming truthy, so it resets before the modal shows...

**Reply by Tim-Wils** [reply]
Maybe it could just be firing a CustomEvent named something like 'x-transition-end' that dispatches at https://github.com/alpinejs/alpine/blob/main/packages/alpinejs/src/directives/x-transition.js#L251
That way we can all use it like this: `<div x-transition x-show="shown" x-on:x-transitionend="shown ? doSomethingAfterTransitionShown() : doSomethingAfterTransitionHide()">`

**Reply by nikitaourazbaev** [reply]
@ekwoka Ah, that’s a nice tip about `destroy`, I didn’t know that, thanks! Does that work for other data contexts too?

> Is there a real issue if you do the teardown logic just upon the variable becoming falsy?

Yes, the modal becomes blank before it has transitioned out.

For each modal, there are two variables: `open`, which is a boolean and `content`, which is a string key. 

Each modal contains a “slot”, which is just a div with a unique id. Lots of other places throughout the code can `x-teleport` into that slot out of an `x-if` that checks for the `content` key string. This allows us to have a few modal slots that can show all kinds of different content on demand.

So the close method sets both `open` to `false` and `content` to `''`. For all modals until now we could blank out the content in the open method, before changing it to something else. But for this specific one we need to blank it out upon close (I’m actually not sure why, I’ll check with my team mate as I didn’t work on this specific case – maybe we can refactor to blank it out on open). So in this case we run the close method, and the modal becomes blank right away, then transitions out. Putting the content reset into an `_x_doHide` override does solve this pretty well for now.

While I can’t remember specifics now, I do know I’ve run into wanting to know whether something has been hidden or not a few times in the past, too.

@Tim-Wils `x-transition-end` could definitely be useful for other cases but I’m more looking for a way to know whether `x-show` has hidden or shown something _regardless_ of whether a transition is used.

**Reply by ekwoka** [reply]
Ah, I'd personally just not do it with teleporting. Just have the modal with the button in a local context.

`destroy`works on all x-data that might get removed.

It's basically the cleanup for the x-data directive, like how x-on removes listeners and such.

**Reply by asbjornu** [reply]
I just had a use for this by invoking `focus()` on an `<input>` within a `<div>` decorated with `x-show`, to focus the `<input>` element immediately after the `<div>` is shown:

```html
<div x-show="…" x-cloak x-transition @transitionend.once="$refs.title.focus()">
    <label>
        Title
        <input x-model="title" type="text" name="title" x-ref="title">
    </label>
</div>
```

It works, which is great. It was just now that I learned that [`transitionend` is a core CSS event](https://developer.mozilla.org/en-US/docs/Web/API/Element/transitionend_event) and not something specific to Alpine.js. Which sort of explains why I couldn't find any documentation about `transitionend` in the Alpine.js documentation. But perhaps a little mention of it and the other `transition…` events on the [`x-transition` directive page](https://alpinejs.dev/directives/transition) would be a good idea?

Alternatively, an Alpine.js-specific `x-on:show` event would be very useful. As @nikitaourazbaev, I'm not necessarily in need of `x-transition` to show the above `<div>`, I just found adding `x-transition` to be the easiest way to have `x-on:transitionend` to give the `<input>` focus work reliably.

Or is there perhaps a completely different way to do this in Alpine.js?

**Reply by nikitaourazbaev** [reply]
@asbjornu I think the [Focus plugin](https://alpinejs.dev/plugins/focus) should work pretty well for this.

**Reply by mynona** [reply]
Yes, that would be great if the x-on:show event could be extended. I agree. Would make much more sense.

@nikitaourazbaev  The Focus plugin itself cannot handle it and I was looking for weeks for a solution and happily I found this work around with transitioned now.

**Reply by ekwoka** [reply]
Alternatively, use x-if and then x-init on the element in question.

**Reply by nikitaourazbaev** [reply]
What I ended up doing in one of the cases where I needed this, is awaiting a promise of all animations’ `finished` promises from `getAnimations()` inside a try / catch statement. This will wait until all transitions on an element are completed, and will work whether the element has a transition or not. 

I haven’t shared this because I think it’s pretty specific to my use case but maybe this will help someone. 

You need to already know that the transition will happen (maybe with a `$watch` on the thing that causes the `x-show` to toggle or with an event), and you need to have a reference to the element with the `x-show` element that is being shown / hidden (here it’s `el`).

```js
try {
  await Promise.all(
    el.getAnimations().map((animation) => animation.finished)
  );

  runOnOpen();
} catch (e) {
  runOnOpen();
}
```

**Comment by ludwig-gramberg** [comment]
I had the same problem. I needed to know when all content panels in an accordion had finished changing their visibility (changed by x-show) in order to scroll to the currently open one.
I ended up using a mutation observer and making the elements distinguishable by a custom attribute. 
this will work as long as alpine does not change how it applies x-show.

```js
// this is code is run inside the init-function of the component

const debounceScroll = window.Alpine.debounce(() => {  
    
    myScrollMethod()
    
}, 100)
const observer = new MutationObserver(mutations => {
    mutations.forEach(mutation => {
        if( mutation.type === 'attributes' && 
            mutation.attributeName === 'style' &&
            mutation.target.hasAttribute('data-accordion-content')
        ) {
            debounceScroll()
        }
    })
})
observer.observe(this.$root, {
    attributeFilter: ['style'],
    subtree: true,
})
```

**Comment by nikitaourazbaev** [comment]
I ended up making [a small plugin](https://www.npmjs.com/package/x-shown) to hopefully help with this in some cases. It adds `x-shown` and `x-hidden` directives, to be used on the same element that has `x-show`.
